file-upload/file-upload-scala-sbt/fileupload-impl/src/main/scala/com/example/fileupload/impl/FileuploadServiceImpl.scala

```diff
@@ -0,0 +1,120 @@
+                             playBodyParsers: PlayBodyParsers
+                           )(implicit mat: Materializer,
+                             exCtx: ExecutionContext
+                           ) extends FileuploadService {
+
+  override def uppercaseEcho() = ServiceCall { input =>
+    Future.successful(input.toUpperCase)
+  }
+
+
+  override def uploadFile(): ServiceCall[NotUsed, String] = PlayServiceCall {
+    callback: (ServiceCall[NotUsed, String] => EssentialAction) =>
+
+      // (1) Create a multi-part body parser with the Accumulator (see the implementation of 'fileHandler()' below)
+      //
+      // Play provides a varied set of body parsers. In our case we need 'multipartFormData'
+      // but might also use 'json', 'raw', ... . We complete the build of the bodyParser providing
+      // the fileHandler we created below.
+      val bodyParser: BodyParser[MultipartFormData[FilePath]] = playBodyParsers.multipartFormData(fileHandler)
+
+      // (2) Create Play's Action.
+      //
+      // Play handles incoming request via Action (aka EssentialAction). Lagom is built on top of
+      // Play and automatically creates an Action for each ServiceCall. But, when implementing
+      // a ServiceCall using a PlayServiceCall like we disable the automatic creation and so we
+      // have to create the Play Action ourselves. The callback argument above allows converting
+      // a regular ServiceCall into an Action so that can mix Pal and Lagom coding styles.
+      //
+      // So, we create an 'async' action with a multi-part body parser that uses our own file handling.
+      // The missing bit on an `actionBuilder.async(bodyParser)` is a function
+      // `Request => Future[Result]` (that is regular Play API). In this example we create
+      // a requestHandling(callback)(request) method which given a `Request` provides the `Future[String]`
+      //
+      // Note: Building a `requestHandling(...)` and using the provided `callback` is not a required step
+      // and you could implement all your ServiceCall using only Play features.
+      actionBuilder.async(bodyParser)(requestHandling(callback))
+  }
+
+  // (2.extra) Using the callback provided by the lagom framework we can implement
+  // our logic using a regular Lagom ServiceCall. The provided `request` will contain the
+  // information generated in the custom body parser (in our case a `FilePath`).
+  private def requestHandling(
+                               callback: ServiceCall[NotUsed, String] => EssentialAction
+                             )(
+                               request: Request[MultipartFormData[FilePath]]
+                             ): Future[Result] = {
+    val wrappedAction: EssentialAction = callback(
+      ServiceCall { notused =>
+        val files: Seq[FilePart[FilePath]] = request.body.files
+        Future.successful(files.map{_.ref.absolutePath}mkString("Uploaded[", ", ", "]"))
+      })
+    wrappedAction(request).run()
+  }
```

Comment:
{{nitpicking}}
Can we reformat this part? it's really looking weird. 
Also the `map{...}mkString` looks strange and hard to read.
{{nitpicking}}

Reply:
this is gone in the new rewrite